Changes AI holopad functionality

AlinaStarkova opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
its made to be much more functional and better quality usage

all changes only exist for the AI

1. opening the holopad GUI is now on left click, to be in line with every machine
2. when an AI examines a holopad it now shows the AI specific controls (need to shift+right click and examine from the menu, cant shift-click examine, issue with how the AI itself functions)
3. right click will now instantly project a hologram, no more having to click to center your camera, then click again to project
4. your last location is stored as a var when you Project from a chat request for a holocall, so when you then right click any holopad again to un-project it will set your camera back to where you originally started (NOTE: if you simply move far enough away from a holopad to cancel projection you do not move back, you can also ctrl-leftclick to cancel projection and not move back)
5. when someone requests an AI prescence, the chat message the AI gets has two clickables, the first is how it used to be: if you click the location button, you simply move your camera to the holopad without projecting, useful to scout out first, this also sets your last location for hitting ~ to move back, if you however click the "Project Hologram?" button you will immediately project a hologram, this ties in to change 4, if you project first, and then right click the pad, you immediately move back to your last location
6. Ctrl+leftclicking on the holopad will end the holopad call directly without jumping you back to your last location (it does nothing if you are not already a hologram)
7. hitting ~ or 0 will jump you back to your last location if you clicked on someones name in chat or jumped location

for Example: same SM is delamming and your watching over it, your presence is requested in idk, bar, you can project hologram, and when your done you can right click the pad and your immediately back over the SM, so you can now answer holopad requests without losing your position

link to the video to see (its too big to upload here so this is the link to it on the discord)
(a bit out of date but does show part of what i changed)
https://media.discordapp.net/attachments/326831214667235328/864652406583459850/AI-holopad.mp4
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
better functionality and adds better quality usability
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->
better functionality and adds better quality usability

## Changelog
:cl: Nari Harimoto
qol: AI now opens the holopad with left click, and projects a hologram with right click. AI's right clicking the holopad will immediately project, no more double clicking
add: AI's right clicking any holopad while projected via the "Project Hologram?" button when presence is requested, will cancel the projection and put their camera back to where they started (moving far away from the pad will not do this, nor would ctrl-clicking the pad)
add: when an AI's presence is requested to a holopad they get a chat message that has two options, jump to the location, works as it does currently, and "Project Hologram" which will immediately project a hologram there, right clicking the pad will move you back to where you were
add: Ctrl+leftclicking on the holopad will end the holopad call directly without jumping you back to your last location (it does nothing if you are not already a hologram)
qol: AI's can now examine a holopad to see the AI specific controls for the pads (need to shift+right click and examine from the menu, cant shift-click examine, issue with how the AI itself functions.)
qol: when you jump to someone from their name in chat, or a location from chat such as when presence is requested, you can hit "~" or "0" to jump back
qol: when you jump back to your previous location with the "Jump Back" key or to a saved location with "1-9" it stops all following, so you can jump back or jump to a location and you wont jump back to the person you tracked
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

AlinaStarkova committed:
Changes AI holopad functionality

AlinaStarkova committed:
added a forgotten control to the examine

Wayland-Smithy commented on code/game/machinery/hologram.dm:
```diff
@@ -172,7 +172,9 @@ Possible to do for anyone motivated enough:
 	. = ..()
-	if(in_range(user, src) || isobserver(user))
+	if(in_range(user, src) && isAI(user))
+		. += span_notice("The status display reads: Current projection range: <b>[holo_range]</b> units. Use :h to speak through the projection. Right-click to project or cancel a projection. Alt-click to hangup all active calls.")
```
```suggestion
		. += span_notice("The status display reads: Current projection range: <b>[holo_range]</b> units. Use :h to speak through the projection. Right-click to project or cancel a projection. Alt-click to hangup all active and incoming calls.")
```

AlinaStarkova committed:
removed the range check for the AI examine, AI should have infinite range for this

AlinaStarkova committed:
made the examine more accurate

AlinaStarkova committed:
swapped return values

Wayland said it should be this way around, it doesnt appear to make any difference whatsoever in testing so may as well do as suggested

zxaber commented:
The return function is probably unnecessary. We have a camera hotkey function for AIs (ctrl+number to save a location, number alone to jump), which players can use if they want a quick return.

Actually, we have a goto/return setup with hotkeys and the cam_prev var, which is set when you use a camera hotkey to jump to a location (so you can press ~ to jump back). So you could just update that when the player makes a hologram, and then leave it up to the player if they want to return after ending the call.

AlinaStarkova commented:
> 
> 
> The return function is probably unnecessary. We have a camera hotkey function for AIs (ctrl+number to save a location, number alone to jump), which players can use if they want a quick return.
> 
> Actually, we have a goto/return setup with hotkeys and the cam_prev var, which is set when you use a camera hotkey to jump to a location (so you can press ~ to jump back). So you could just update that when the player makes a hologram, and then leave it up to the player if they want to return after ending the call.
> 
> Otherwise, good QoL changes.

the return hotkey ONLY applies for jumps, it does not apply for moving around while projecting, nor would it work if you used the new instant project button when your requested to a holopad. as such the return function is a neseccity to compensate for the reduced functionality of the return hotkey when projecting

zxaber commented:
>nor would it work if you used the new instant project button when your requested to a holopad.

I mean, it *could*, if you set it in the proc that runs when you click the instant project link. That was the suggestion. Y'know, maybe the AI wants to disconnect from the holopad but keep watching the same location or something.

The return button (~) just sends you to the turf saved in the cam_prev var, so all you'd have to do is `cam_prev = get_turf(eyemob)` before shooting the AI's eye over.

AlinaStarkova commented:
> 
> 
> > nor would it work if you used the new instant project button when your requested to a holopad.
> 
> I mean, it _could_, if you set it in the proc that runs when you click the instant project link. That was the suggestion. Y'know, maybe the AI wants to disconnect from the holopad but keep watching the same location or something.
> 
> The return button (~) just sends you to the turf saved in the cam_prev var, so all you'd have to do is `cam_prev = get_turf(eyemob)` before shooting the AI's eye over.

im not sure the return button is even well known, i dont often see AI mains use that button. and im not sure doing it that way is entirely needed, the majority of holopads are too far from each other, just moving away from the pad keeps your position, in fact i dont think most AIs even disconnect from the hologram explicitly by clicking the pad to begin with.

i can look into changing it to the method you suggest but im worried that it diminishes the qol of change due to obscurity of the ~ hotkey, and potential conflicts with that var. what if you project, then double click nearby without breaking your projection (double clicking as an AI moves your camera iirc) and that resets that cam_last var on you? just to name one example of how the var could get reset on you

zxaber commented:
For what it's worth, the cam_prev variable currently is only set by camera location hotkeys. I've been considering having other jumps (like clicking someone's name in chat) also set it, but I haven't made such a change yet.

Honestly, it's probably not anything worth like losing sleep over. If the first link that just jumps you to the location of the holopad doesn't set up an auto return (and I assume it doesn't), then it's whatever. I *personally* would feel the annoyance of being jumped back without my control or input after ending a holocall, but then I'll probably want to see the room before spawning my hologram anyway.

AlinaStarkova commented:
>  but then I'll probably want to see the room before spawning my hologram anyway.

thats why i left the original option, the second option is meant more for AIs who are busy but wish to quickly get the call over with. oftentimes as an AI i will ignore a request because im actively engaged in a law 1 situation (like presiding over an attack, stopping an SM delam etc..) but this instant project will let me quickly talk then jump back to where i was smoothly without worrying about hotkeys (or remembering the jump back hotkey even exists, im still convinced most people dont know it exists)

i suppose i could potentially make it so if your a hologram, left clicking a pad hangs up the call without jumping you back, rather then opening the UI (an AI probably has no need to open a holopad UI while projected) or mayble ctrl+right click will do that idk

AlinaStarkova commented:
> 
> 
> For what it's worth, the cam_prev variable currently is only set by camera location hotkeys. I've been considering having other jumps (like clicking someone's name in chat) also set it, but I haven't made such a change yet.

this also probably wont work very well, when you select someones name you enter an orbit of them of sorts, if you try to move your camera via jumping you just jump back to the person your orbiting, so to do this the last jump button needs to cancel the orbit before jumping back

AlinaStarkova committed:
Ctrl-left-click as a hologram ends the call without jumping back

AlinaStarkova committed:
Updated, some changes

made the ~ key jump back when jumping via hotkeys or jumping to people in chat

AlinaStarkova committed:
Changes requested

Kylerace commented on code/game/machinery/hologram.dm:
```diff
@@ -345,25 +347,34 @@ Possible to do for anyone motivated enough:
-	else//If there is a hologram, remove it.
+	else//If there is a hologram, remove it, and jump to your last location.
+		if(user.lastloc)//only jump to your last location if your lastloc is set, which only sets if you projected from a request message.
+			SetLoc(user.eyeobj) = user.lastloc
```
```suggestion
			user.eyeobj.SetLoc(user.lastloc)
```

Kylerace commented on code/game/machinery/hologram.dm:
```diff
@@ -403,14 +414,15 @@ Possible to do for anyone motivated enough:
 		var/obj/effect/overlay/holo_pad_hologram/Hologram = new(loc)//Spawn a blank effect at the location.
 		if(AI)
 			Hologram.icon = AI.holo_icon
+			SetLoc(AI.eyeobj) = get_turf(src) //ensure the AI camera moves to the holopad
```
```suggestion
			AI.eyeobj.SetLoc(get_turf(src)) //ensure the AI camera moves to the holopad
```

AlinaStarkova committed:
removed a redundant src

AlinaStarkova committed:
Merge branch 'master' into AI-holopad

AlinaStarkova committed:
fixed linters
